### PR TITLE
support for Vue 2.6

### PR DESCRIPTION
**Pull request by farfromrefug**

### CODE REVIEW

**Comment by tralves** [comment]
Hi @farfromrefug! Awesome work! This PR looks ok and I need 2.6 in my project, but there are lots of changes here.
Can you please describe in more detail what you did and why? I think it will help @rigor789 review and accept it.

@rigor789, please review and accept it 😛  I am going to test this on my projects to see if I find any issues.

**Comment by farfromrefug** [comment]
Basically i did 3 things:
- fixed the build process. It was broken and the ```nativescript-vue-template-compiler``` was not building anymore. Mostly this was due to ```registerElement``` being needed within webpack. Because of that when doing ```require('nativescript-vue-template-compiler')``` within your webpack config you were trying to import ```tns-core-modules``` which can't work. So i had to refactor things a bit and create a different elementRegistry for webpack/runtime (though they can actually be merge, i know how to do it).
- update the runtime to work with vue 2.6. It broke a few things within the ```nativescript-vue``` and i am not even sure i have seen them all yet. For now i saw that the view directive and the listview patchTemplate were broken.
- remove the console.log override as it was creating crash within my apps with complex object logging

Already using that branch within my project without error

**Comment by rigor789** [comment]
@tralves @farfromrefug it's on my todo list to go over this as soon as I can - but I can't promise anything...

File: build/config.js [hunk 1]
```diff
@@ -57,6 +64,8 @@ const genConfig = (name) => {
         input: opts.entry,
         external: opts.external,
         output: {
+            strict: false,
```
Comment (rigor789) [inline build/config.js]: @farfromrefug Can you explain why this is required? 

According to rollup docs the `strict` option does the following:
> Whether to include the 'use strict' pragma at the top of generated non-ESM bundles. Strictly-speaking, ES modules are always in strict mode, so **you shouldn't disable this without good reason**.

File: build/config.js [hunk 2]
```diff
@@ -17,6 +17,12 @@ const banner = (name, version) => `
  * Released under the MIT license.
  */
 `
+const intro = `
+if (!this['process']) {
```
Comment (rigor789) [inline build/config.js]: @farfromrefug any reason you are not checking for `global.process`? In this case I believe it does the same thing, since we are in a global context - but still curious if there is something I'm not aware of! :)

File: build/config.js [hunk 3]
```diff
@@ -72,10 +81,9 @@ const genConfig = (name) => {
             replace({
                 __WEEX__: false,
                 __VERSION__: VueVersion,
-                'process.env.NODE_ENV': "'development'",
```
Comment (rigor789) [inline build/config.js]: @farfromrefug Does this remove all the development related help messages that vue prints? If so - can this be re-enabled while developing?

File: platform/nativescript/compiler/modules/class.js [hunk 4]
```diff
@@ -4,17 +4,17 @@ import { parseText } from 'compiler/parser/text-parser'
...
-          'Use v-bind or the colon shorthand instead. For example, ' +
-          'instead of <div class="{{ val }}">, use <div :class="val">.'
-      )
-    }
-  }
+  // if (process.env.NODE_ENV !== 'production' && staticClass) {
+  //   const expression = parseText(staticClass, options.delimiters)
+  //   if (expression) {
+  //     warn(
+  //       `class="${staticClass}": ` +
```
Comment (rigor789) [inline platform/nativescript/compiler/modules/class.js]: @farfromrefug any reason this has been removed? Interpolation in attributes should not be used - and should be warned about

File: platform/nativescript/compiler/modules/view.js [hunk 5]
```diff
@@ -11,8 +11,8 @@ function transformNode(el) {
       mods[mod] = true
       return mods
     }, {})
-    getAndRemoveAttr(el, attr)
-    addDirective(el, 'view', `v-view:${attrName}`, '', arg, modifiers)
+    getAndRemoveAttr(el, attr, true)
+    addDirective(el, 'view', `v-view:${attrName}`, '', false, arg, modifiers)
```
Comment (rigor789) [inline platform/nativescript/compiler/modules/view.js]: @farfromrefug what do the added args do?  

The first one is `removeFromMap` - why is this required?

The second one seems to be mapping to `arg` - is this the intended parameter? Or should that be `isDynamicArg` instead?

```js
export function addDirective (
  el: ASTElement,
  name: string,
  rawName: string,
  value: string,
  arg: ?string,
  isDynamicArg: boolean,
  modifiers: ?ASTModifiers,
  range?: Range
)
```

File: platform/nativescript/framework.js [hunk 6]
```diff
@@ -19,27 +13,6 @@ setVue(Vue)
 Vue.use(ModalPlugin)
 Vue.use(NavigatorPlugin)
 
-const newLineRegExp = /\\n/g
-
-console.log = (function(log, inspect, Vue) {
-  return function(...args) {
```
Comment (rigor789) [inline platform/nativescript/framework.js]: @farfromrefug I'm not too keen on removing this - I'd rather make it opt-out, as I've had a much better development experience with this in place vs. the default console.log.

File: platform/nativescript/framework.js [hunk 7]
```diff
@@ -19,27 +13,6 @@ setVue(Vue)
...
-          colors: Vue.config.debug,
-          showHidden: true
-        }).replace(newLineRegExp, '\n')
-      })
-    )
-  }
-})(console.log, inspect, Vue)
-
-console.keys = function(object) {
-  console.log(Object.keys(object))
```
Comment (rigor789) [inline platform/nativescript/framework.js]: @farfromrefug same with this one - there is no default `console.keys` as far as I'm aware in NativeScript (unless it has been added recently) - so this shouldn't be a problem, but can be made opt-out (under the same umbrella as console.log overrides)

File: platform/nativescript/runtime/directives/view.js [hunk 8]
```diff
@@ -1,7 +1,8 @@
 export default {
-  inserted(el, { arg, modifiers }) {
+  inserted(el, binding, vnode, oldNode) {
     const parent = el.parentNode.nativeView
-
+    const arg = binding.rawName.split(":")[1];
+    const modifiers = binding.modifiers;
```
Comment (rigor789) [inline platform/nativescript/runtime/directives/view.js]: @farfromrefug I'm assuming this is a related change to `platform/nativescript/compiler/modules/view.js`, can you elaborate why this is needed?

File: platform/nativescript/util/index.js [hunk 9]
```diff
@@ -1,6 +1,6 @@
-import { isKnownView, getViewMeta, getViewClass } from '../element-registry'
+import { isKnownView, getViewMeta, getViewClass } from '../element-register'
 import { makeMap, once } from 'shared/util'
-import { VUE_VM_REF } from '../runtime'
+ const VUE_VM_REF = '__vue_vm_ref__'
```
Comment (rigor789) [inline platform/nativescript/util/index.js]: @farfromrefug Why do we need to re-declare rather than importing? Is this due to the fact that `runtime/index` relies on `tns-core-modules` and should not be referenced to avoid build issues?

Comment (farfromrefug) [reply platform/nativescript/util/index.js]: @rigor789 yes indeed. May be we should have a ```constants.ts``` with the constant define ?

File: platform/nativescript/runtime/directives/view.js (hunk 8 above)
Comment (farfromrefug) [reply platform/nativescript/runtime/directives/view.js]: it seems to have changed in Vue 2.6, your was not working anymore and i could not get the "arg" as before

File: platform/nativescript/framework.js (hunk 6 above)
Comment (farfromrefug) [reply platform/nativescript/framework.js]: @rigor789 ok i understand. Make it an opt-in then. GDPR ;) It still crashes for me + i don't really want it. And users won't understand why. If you can't recall why it was necessary then i think it really should be opt-in and not opt-out

File: build/config.js (hunk 2 above)
Comment (farfromrefug) [reply build/config.js]: Well to be honest i am not sure how "process" is handled in Nativescript. What i know is that VueJs checks ```process``` and not ```global.process```. That's why i used this.

File: build/config.js (hunk 1 above)
Comment (farfromrefug) [reply build/config.js]: I had some issues with Vue and Nativescript in strict mode and {N} guys told me to remove strict. I did it while i was testing. Will remove it

File: platform/nativescript/compiler/modules/class.js (hunk 4 above)
Comment (farfromrefug) [reply platform/nativescript/compiler/modules/class.js]: Well it was a deprecation warning that was implying a regexp test on every transform node. Seemed like it was time to remove it. But i can bring it back if you prefer

File: platform/nativescript/compiler/modules/view.js (hunk 5 above)
Comment (farfromrefug) [reply platform/nativescript/compiler/modules/view.js]: oh my bad i inversed the args. Maybe that's why arg was not working anymore! will fix

File: build/config.js (hunk 2 above)
Comment (rigor789) [reply build/config.js]: @farfromrefug `process` is not defined in NativeScript by default, and the reason we're adding it manually is that some vue plugins check for `process.env` which then causes issues since it's trying to access property `env` of undefined. Checking if it exists on `global` and then adding to it ensures that both `global.process` and just `process` are defined.

File: platform/nativescript/framework.js (hunk 7 above)
Comment (farfromrefug) [reply platform/nativescript/framework.js]: @rigor789 again should be opt-in. Especially as you dont have typings for this, right?

File: platform/nativescript/compiler/modules/class.js (hunk 4 above)
Comment (rigor789) [reply platform/nativescript/compiler/modules/class.js]: @farfromrefug Just checked and it is still present in vue (web), so let's keep it there until even vue removes it! :)

https://github.com/vuejs/vue/blob/ebc1893faccd1a9d953a8e8feddcb49cf1b9004d/src/platforms/web/compiler/modules/class.js#L16-L22

File: platform/nativescript/framework.js (hunk 6 above)
Comment (rigor789) [reply platform/nativescript/framework.js]: @farfromrefug the reason I had to add it was that the built-in console.log crashed the app for anything more "complex" than a string, or a simple object `{ foo: "bar" }`. Most notably it crashed whenever trying to log a vue observed property, which was probably 90% of the things you'd log.

Plus it adds colors - which I really like when looking at the logs, makes it a lot easier to find things (this might be just my preference, and it's disabled by default on the playground anyways)

File: platform/nativescript/util/index.js (hunk 9 above)
Comment (rigor789) [reply platform/nativescript/util/index.js]: @farfromrefug yes, I like the idea of a constants file - should keep the codebase tidier!

File: build/config.js (hunk 2 above)
Comment (farfromrefug) [reply build/config.js]: @rigor789 VueJs 2.6 also uses it. So you are saying that i should do?:
```
if (!global['process']) {
  global.process = process = {env:{}}
}
```

File: build/config.js (hunk 3 above)
Comment (farfromrefug) [reply build/config.js]: well it s the user choice by replacing within his weback config. that code was making the appear always, even in prod

File: platform/nativescript/runtime/directives/view.js (hunk 8 above)
Comment (rigor789) [reply platform/nativescript/runtime/directives/view.js]: @farfromrefug I'm assuming it's this commit that changed directives: https://github.com/vuejs/vue/commit/dbc0582587f90e78867809bb6ae683301cd0626b

wonder if changing the order in `platform/nativescript/compiler/modules/view.js` for the arguments would make it work without this change.

File: build/config.js (hunk 2 above)
Comment (rigor789) [reply build/config.js]: @farfromrefug yeah that looks good to me, just change `this['process']` to `global['process']`

File: build/config.js (hunk 3 above)
Comment (rigor789) [reply build/config.js]: @farfromrefug alright, so as far as I understand this change, this should still allow the help messages to appear unless webpack sets the env to "production"?

File: platform/nativescript/compiler/modules/view.js (hunk 5 above)
Comment (rigor789) [reply platform/nativescript/compiler/modules/view.js]: @farfromrefug perfect, hopefully that will fix it!

File: platform/nativescript/framework.js (hunk 7 above)
Comment (rigor789) [reply platform/nativescript/framework.js]: @farfromrefug that's correct, there are no typings. I was pretty sure this was a standard method on console - but apparently not. In that case I'm happy to remove this one - and move  it to user-land if needed.

File: platform/nativescript/runtime/directives/view.js (hunk 8 above)
Comment (farfromrefug) [reply platform/nativescript/runtime/directives/view.js]: yes will try

File: platform/nativescript/framework.js (hunk 6 above)
Comment (farfromrefug) [reply platform/nativescript/framework.js]: @rigor789 those crashes do not appear anymore. Plus the contrary happens. Crashes with it.
If you use it just for the colors, should be in userland. plus it adds unecessary dependencies in production.

File: build/config.js (hunk 3 above)
Comment (farfromrefug) [reply build/config.js]: @rigor789 indeed it will appear by default because it checks for ```NODE_ENV !== 'production'``` which is the case as ```NODE_ENV``` is not defined. 
However that change allow the user tu define ```NODE_ENV``` to ```production``` when building for production, and thus remove all the warning. 
It is just moved to userland

File: platform/nativescript/util/index.js (hunk 9 above)
Comment (farfromrefug) [reply platform/nativescript/util/index.js]: done

File: platform/nativescript/runtime/directives/view.js (hunk 8 above)
Comment (farfromrefug) [reply platform/nativescript/runtime/directives/view.js]: fix working as we talked about

File: platform/nativescript/framework.js (hunk 6 above)
Comment (farfromrefug) [reply platform/nativescript/framework.js]: @rigor789 so we can keep this in userland?

File: platform/nativescript/compiler/modules/view.js (hunk 5 above)
Comment (farfromrefug) [reply platform/nativescript/compiler/modules/view.js]: done

File: platform/nativescript/compiler/modules/class.js (hunk 4 above)
Comment (farfromrefug) [reply platform/nativescript/compiler/modules/class.js]: fixed

File: build/config.js (hunk 2 above)
Comment (farfromrefug) [reply build/config.js]: fixed, testing ```global.process```

**Comment by tralves** [comment]
I think I found another issue.
This branch doesn't seem to support the new scope slot syntax `v-slot`. I thought it would support them automatically, but it seems like ns-vue needs to do something extra. See  [platform/nativescript/compiler/modules/scopedSlots.js](https://github.com/nativescript-vue/nativescript-vue/blob/8f3593b3d673efda81d6d3895dc0706b344d0595/platform/nativescript/compiler/modules/scopedSlots.js).

**Comment by tralves** [comment]
I think I found another issue.
This branch doesn't seem to support the new scope slot syntax `v-slot` introduced in 2.6...

**Comment by farfromrefug** [comment]
> I think I found another issue.
> This branch doesn't seem to support the new scope slot syntax `v-slot` introduced in 2.6...
> 
> I get this: `[Vue warn]: Failed to resolve directive: slot`
> I thought maybe NS-Vue is doing something weird with directives?...

Is is what you use with ```nativescript-multi-drawer``` ? Cause if it is, it's working for me. Can you share an example project?

**Comment by tralves** [comment]
This is the example project: https://github.com/tralves/ns-vue-v-slot-test

This is the relevant part in `App.Vue`:

```xml
<StackLayout>
  <Label class="message" text="OLD SYNTAX (slot-scope)" col="0" row="0"/>
  <CurrentUser>
    <template slot-scope="scope">{{scope.user.firstName}} {{scope.user.lastName}}</template>
  </CurrentUser>

  <Label class="message" text="NEW SYNTAX (v-slot)" col="0" row="0"/>
  <CurrentUser>
    <template v-slot="scope">{{ scope ? scope.user.firstName : '(scope is not defined)'}}</template>
  </CurrentUser>
</StackLayout>
```

The old syntax `slot-scope` works, but the new syntax `v-slot` doesn't (see the [docs](https://vuejs.org/v2/guide/components-slots.html#Scoped-Slots)).

That warn `[Vue warn]: Failed to resolve directive: slot` is not showing anymore. I think I was not using it with `<template>`...

**Comment by farfromrefug** [comment]
@tralves i see. Don't use that feature yet. Think it should be good to merge that PR first. Then implement that new feature. @rigor789 any news on this?
I need to know if you are about to merge this. Otherwise i will make my fork public under another name.

Thanks

**Comment by rigor789** [comment]
@farfromrefug I will try to go over it in the next few days!

**Comment by keithgulbro** [comment]
Hi @tralves,

Any update on the use of v-slot? I'm attempting to use named slots and still encountering an issue, I think this feature is of significant value given VueJS is designed to use component tree's.

**Comment by tralves** [comment]
@keithgulbro no news here. I tried to find the cause but couldn't figure it out. But you can still use the [deprecated slot syntax](https://vuejs.org/v2/guide/components-slots.html#Deprecated-Syntax). That one works.

**Comment by farfromrefug** [comment]
@tralves think i just found why it is not working with the new syntax.
Will try and fix it. 
But i think it will be in a fork.

**Comment by rigor789** [comment]
Changes landed in #487 and [v2.3.0-rc.0](https://github.com/nativescript-vue/nativescript-vue/releases/tag/v2.3.0-rc.0)

**Comment by farfromrefug** [comment]
@rigor789 will try your rc. Thanks
